dbms/src/AggregateFunctions/AggregateFunctionMinMaxWindow.h

```diff
@@ -0,0 +1,393 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <AggregateFunctions/AggregateFunctionMinMaxAny.h>
+#include <AggregateFunctions/IAggregateFunction.h>
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnVector.h>
+#include <Common/typeid_cast.h>
+#include <DataTypes/IDataType.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteHelpers.h>
+#include <common/StringRef.h>
+
+namespace DB
+{
+template <typename T>
+struct SingleValueDataFixedForWindow : public SingleValueDataFixed<T>
+{
+private:
+    using Self = SingleValueDataFixedForWindow<T>;
+    using ColumnType = std::conditional_t<IsDecimal<T>, ColumnDecimal<T>, ColumnVector<T>>;
+
+    mutable std::deque<T> * saved_values;
+
+public:
+    SingleValueDataFixedForWindow()
+        : saved_values(nullptr)
+    {}
+
+    ~SingleValueDataFixedForWindow() { delete saved_values; }
+
+    void insertMaxResultInto(IColumn & to) const { insertMinOrMaxResultInto<false>(to); }
+
+    void insertMinResultInto(IColumn & to) const { insertMinOrMaxResultInto<true>(to); }
+
+    template <bool is_min>
+    void insertMinOrMaxResultInto(IColumn & to) const
+    {
+        if (this->has())
+        {
+            auto size = saved_values->size();
+            T tmp = (*saved_values)[0];
+            for (size_t i = 1; i < size; i++)
+            {
+                if constexpr (is_min)
+                {
+                    if ((*saved_values)[i] < tmp)
+                        tmp = (*saved_values)[i];
+                }
+                else
+                {
+                    if (tmp < (*saved_values)[i])
+                        tmp = (*saved_values)[i];
+                }
+            }
+            static_cast<ColumnType &>(to).getData().push_back(tmp);
+        }
+        else
+        {
+            static_cast<ColumnType &>(to).insertDefault();
+        }
+    }
+
+    void prepareWindow() { saved_values = new std::deque<T>(); }
+
+    void reset()
+    {
+        this->has_value = false;
+        saved_values->clear();
+    }
+
+    void decrease()
+    {
+        saved_values->pop_front();
+        if unlikely (saved_values->empty())
+            this->has_value = false;
+    }
+
+    bool changeIfLess(const IColumn & column, size_t row_num, Arena * arena)
+    {
+        auto to_value = static_cast<const ColumnType &>(column).getData()[row_num];
+        if (saved_values != nullptr)
+            saved_values->push_back(to_value);
+
+        return SingleValueDataFixed<T>::changeIfLess(column, row_num, arena);
+    }
+
+    bool changeIfLess(const Self & to, Arena * arena)
+    {
+        if (saved_values != nullptr)
```

Comment:
Is this necessary ? you mean `prepareWindow()` is not always called?
but `reset()` didn't check if `saved_values` is nullptr

Reply:
> Is this necessary ? you mean `prepareWindow()` is not always called? but `reset()` didn't check if `saved_values` is nullptr

I think the check is needless, I will delete it.

Reply:
> Is this necessary ? you mean `prepareWindow()` is not always called? but `reset()` didn't check if `saved_values` is nullptr

Yes, in some circumstance, we will never call decrease function.